packages/core/src/render3/di.ts

```diff
@@ -344,6 +344,100 @@ export function injectAttributeImpl(tNode: TNode, attrNameToInject: string): str
   return null;
 }
 
+/**
+ * Returns a `TNode` of the location where the current `LView` is inserted into.
+ *
+ * @param lView an `LView` that we want to find parent `TNode` for.
+ */
+function getParentTNode(lView: LView): TNode|null {
+  const tView = lView[TVIEW];
+  if (tView.type === TViewType.Embedded) {
+    ngDevMode && assertDefined(tView.declTNode, 'Embedded TNodes should have declaration parents.');
+    return tView.declTNode;
+  } else if (tView.type === TViewType.Component) {
+    // Components don't have `TView.declTNode` because each instance of component could be
+    // inserted in different location, hence `TView.declTNode` is meaningless.
+    // Falling back to `T_HOST` in case we cross component boundary.
+    return lView[T_HOST];
+  }
+  // Remaining TNode type is `TViewType.Root` which doesn't have a parent TNode.
+  return null;
+}
+
+/**
+ * Returns the parent TNode and matching LView for a given TNode.
+ *
+ * Either returns the immediate node host or recursively navigates up the tree
+ * until it finds the next valid parent node, if it exists, and returns that.
+ *
+ * @param tNode a `TNode` that we want to find the parent of.
+ * @param lView an `LView` that corresponds to the tNode param.
+ * @param flags an `InjectFlags` that we check to skip @Host.
+ * @returns a tuple of the parent tNode and the corresponding lView of that node or null
+ */
+function getParentNodeView(
```

Comment:
Since the return value is a tuple, we can probably be more verbose with the function name:
```suggestion
function getParentTNodeAndLView(
```

Reply:
Changed!